texar/modules/decoders/rnn_decoders.py

```diff
@@ -660,6 +723,55 @@ def forward(  # type: ignore
 
         return outputs, final_state, sequence_lengths
 
+    def beam_decode(self,
+                    start_tokens: torch.LongTensor,
+                    end_token: int,
+                    embedding_fn: Callable[[torch.LongTensor], torch.Tensor],
+                    initial_state: AttentionWrapperState,
+                    decode_length: int = 256,
+                    beam_width: int = 5,
+                    length_penalty: float = 0.6):
+
+        def _prepare_beam_search(x):
+            x = torch.unsqueeze(x, dim=1)
+            tile_dims = [1] * x.dim()
+            tile_dims[1] = beam_width
+            x = x.repeat(tile_dims)
+            shape = list(x.size())
+            shape[0] *= shape[1]
+            shape.pop(1)
+            x = x.view(shape)
+            return x
+
+        memory_beam_search = _prepare_beam_search(self.memory)
+        memory_sequence_length_beam_search = _prepare_beam_search(
+            self.memory_sequence_length)
+
+        def _symbols_to_logits_fn(ids, state):
+            ids = ids[:, -1]
+            inputs = embedding_fn(ids)
+            wrapper_outputs, wrapper_state = self._cell(
+                inputs, state, memory_beam_search,
+                memory_sequence_length_beam_search)
+            logits = self._output_layer(wrapper_outputs)
+            return logits, wrapper_state
+
+        outputs, log_prob = beam_search(  # type: ignore
```

Comment:
Why do we need a `# type: ignore` here?

Reply:
`self._vocab_size` could be `None` (https://github.com/asyml/texar-pytorch/blob/master/texar/modules/decoders/decoder_base.py#L92).

Reply:
In general it's much better to use assertions to guide the type checker. This adds almost no overhead anyway.